src/QsCompiler/Transformations/Attributes.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.Quantum.QsCompiler.DataTypes;
+using Microsoft.Quantum.QsCompiler.SyntaxTree;
+
+
+namespace Microsoft.Quantum.QsCompiler.Transformations
+{
+    using CallablePredicate = Func<QsCallable, bool>;
+    using AttributeId = QsNullable<UserDefinedType>;
+    using QsRangeInfo = QsNullable<Tuple<QsPositionInfo, QsPositionInfo>>;
+
+
+    /// <summary>
+    /// Contains tools for building and adding attributes to an existing Q# compilation. 
+    /// </summary>
+    public static class Attributes
+    {
+        private static AttributeId BuildId(QsQualifiedName name) =>
+            AttributeId.NewValue(new UserDefinedType(name.Namespace, name.Name, QsRangeInfo.Null));
+
+        // public static methods
+
+        public static QsDeclarationAttribute BuildAttribute(QsQualifiedName name, TypedExpression arg) =>
+            new QsDeclarationAttribute(BuildId(name), arg, null, QsComments.Empty);
+
+        public static TypedExpression StringArgument(string target) =>
+            SyntaxGenerator.StringLiteral(NonNullable<string>.New(target), ImmutableArray<TypedExpression>.Empty);
+
+        public static QsCompilation AddToCallables(QsCompilation compilation, QsDeclarationAttribute attribute, CallablePredicate predicate = null) =>
+            new AddAttributes(new[] { (attribute, predicate) }).Apply(compilation);
+
+        public static QsCompilation AddToCallables(QsCompilation compilation, params (QsDeclarationAttribute, CallablePredicate)[] attributes) =>
+            new AddAttributes(attributes).Apply(compilation);
+
+        public static QsCompilation AddToCallables(QsCompilation compilation, params QsDeclarationAttribute[] attributes) =>
+            new AddAttributes(attributes.Select(att => (att, (CallablePredicate)null))).Apply(compilation);
+
+        public static QsNamespace AddToCallables(QsNamespace ns, QsDeclarationAttribute attribute, CallablePredicate predicate = null) =>
+            new AddAttributes(new[] { (attribute, predicate) }).Namespaces.OnNamespace(ns);
+
+        public static QsNamespace AddToCallables(QsNamespace ns, params (QsDeclarationAttribute, CallablePredicate)[] attributes) =>
+            new AddAttributes(attributes).Namespaces.OnNamespace(ns);
+
+        public static QsNamespace AddToCallables(QsNamespace ns, params QsDeclarationAttribute[] attributes) =>
+            new AddAttributes(attributes.Select(att => (att, (CallablePredicate)null))).Namespaces.OnNamespace(ns);
```

Comment:
It looks like each of these add attributes based on callable predicates? Based on the usecase at microsoft/qsharp-runtime#253, though, it may help to have some static methods or extension methods for make it easier to add attributes to particular callables, given C# values representing literal inputs to those attributes.

For instance, the following LINQ call is used to add a sequence of attributes each of signature `(String, String)` given an `IDictionary<string, string>`:

```C#
        internal static IAttributeBuilder<T> WithDocumentationAttributesFromDictionary<T>(
            this IAttributeBuilder<T> builder, string attributeName, IDictionary<string, string> items
        ) =>
            items
            .Aggregate(
                builder,
                (acc, item) => acc.WithDocumentationAttribute(
                    attributeName,
                    // The following populates all of the metadata needed for a
                    // Q# literal of type (String, String).
                    new TypedExpression(
                        QsExpressionKind<TypedExpression, Identifier, ResolvedType>.NewValueTuple(
                            ImmutableArray.Create(
                                item.Key.AsLiteralExpression(),
                                item.Value.AsLiteralExpression()
                            )
                        ),
                        ImmutableArray<Tuple<QsQualifiedName, NonNullable<string>, ResolvedType>>.Empty,
                        ResolvedType.New(
                            QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>.NewTupleType(
                                ImmutableArray.Create(
                                    ResolvedType.New(QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>.String),
                                    ResolvedType.New(QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>.String)
                                )
                            )
                        ),
                        new InferredExpressionInformation(false, false),
                        QsNullable<Tuple<QsPositionInfo, QsPositionInfo>>.Null
                    )
                )
            );
```

The vast majority of that expression is constructing a `TypedExpression` from two C# strings, suitable for use in constructing a `QsDeclarationAttribute`, such that new SyntaxGeneration methods may help as well.

Reply:
I added a method to build arguments with arbitrary many strings.